src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -306,106 +375,706 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+  if (d4n_writecache == false){
+    ldpp_dout(save_dpp, 10) << "D4NFilterObject::D4NFilterWriteOp::" << __func__ << "(): calling next iterate" << dendl;
+    return next->prepare(y);
   }
+  else
+    return 0;
+}
 
-  return next->process(std::move(data), offset);
+int D4NFilterWriter::process(bufferlist&& data, uint64_t offset)
+{
+    bufferlist bl = data;
+    off_t bl_len = bl.length();
+    off_t ofs = offset;
+    bool dirty = true;
+    rgw::d4n::CacheBlock block, existing_block;
+    auto creationTime = startTime;
+
+    auto version = obj->get_instance();
+
+ldpp_dout(save_dpp, 10) << "AMIN::" << __func__ << "(): " << __LINE__ << " offest is: "  << offset << dendl;
+ldpp_dout(save_dpp, 10) << "AMIN::" << __func__ << "(): " << __LINE__ << " ofs is: "  << ofs << dendl;
+ldpp_dout(save_dpp, 10) << "AMIN::" << __func__ << "(): " << __LINE__ << " bl_len is: "  << bl_len << dendl;
+    std::string prefix;
+    if (version.empty()) { //for versioned objects, get_oid() returns an oid with versionId added
+      prefix =obj->get_bucket()->get_name() + "_" + obj->get_key().get_oid();
+    } else {
+      prefix = obj->get_bucket()->get_name() + "_" + version + "_" + obj->get_key().get_oid();
+    }
+
+ldpp_dout(save_dpp, 10) << "AMIN::" << __func__ << "(): " << __LINE__ << " prefix is: "  << prefix << dendl;
+    rgw::d4n::BlockDirectory* blockDir = driver->get_block_dir();
+
+    block.hostsList.push_back(blockDir->cct->_conf->rgw_local_cache_address); 
+    block.cacheObj.bucketName = obj->get_bucket()->get_name();
+    block.cacheObj.objName = obj->get_key().get_oid();
+    block.cacheObj.dirty = dirty;
+    existing_block.cacheObj.objName = block.cacheObj.objName;
+    existing_block.cacheObj.bucketName = block.cacheObj.bucketName;
+
+
+    int ret = 0;
+
+      if (d4n_writecache == false){
+	
+        std::string oid = prefix + "_" + std::to_string(ofs)+ "_" + std::to_string(bl_len);
+        block.size = bl.length();
+        block.blockID = ofs;
+        block.dirty = false; //writing to the backend
+    	dirty = false;
+	
+  	ldpp_dout(save_dpp, 10) << "D4NFilterObject::D4NFilterWriteOp::" << __func__ << "(): calling next process" << dendl;
+	ret = next->process(std::move(data), offset);
+	
+        if (ret == 0) {
+          driver->get_policy_driver()->get_cache_policy()->update(save_dpp, oid, ofs, bl.length(), version, dirty, creationTime,  obj->get_bucket()->get_owner(), y);
+
+          if (!blockDir->exist_key(&block, y)) {
+            if (blockDir->set(&block, y) < 0) //should we revert previous steps if this step fails?
+  	      ldpp_dout(save_dpp, 10) << "D4NFilterObject::D4NFilterWriteOp::" << __func__ << "(): BlockDirectory set method failed." << dendl;
+            } else {
+              existing_block.blockID = block.blockID;
+              existing_block.size = block.size;
+  	      existing_block.dirty = block.dirty;
+              if (blockDir->get(&existing_block, y) < 0) {
```

Comment:
Wouldn't setting the `existing_block` data in the above lines (1010, 1011, 1012) and then calling `get` for that same block overwrite those values? This applies on lines 1053-1055 as well.

Reply:
We need blockID and size to get the info from the directory. dirty should be removed.